value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,33 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Created by rharter on 5/1/15.
+ */
+public interface AutoValueExtension {
+
+  public interface Context {
+    ProcessingEnvironment processingEnvironment();
+    String packageName();
+    TypeElement autoValueClass();
+    Map<String, ExecutableElement> properties();
+  }
+
+  public interface GeneratedClass {
+    String className();
+    String source();
+    Collection<String> additionalImports();
+    Collection<ExecutableElement> consumedProperties();
+    Collection<String> additionalInterfaces();
+    Collection<String> additionalCode();
```

Comment:
It sounds like the JavaPoet approach is fairly close to what I had added for the Parcelable support, basically allowing extensions to add code directly to the AutoValue class.  I can definitely appreciate the apprehension about that, I was also initially apprehensive, but my inverted inheritance didn't really offer a solution and it didn't seem like ordering extensions was a good solution. 

Also, I completely agree that the CREATOR static field is an unfortunate approach.  I'd love to see something like an annotation on the class pointing to a factory class.  That's kind of what this extension would provide, only generating the factory.  Unfortunately, other [solutions](https://github.com/johncarl81/parceler) require wrapping, which I'm hoping this can avoid.

Regarding the inverted inheritance (AutoValue_Foo (from AutoValue) inherits from $AutoValue_Foo (from extension) inherits from abstract Foo), I don't think that causes problems for the toString use case, provided the AutoValue processor waits until after the extensions are processed to determine methods (toString, equals, hashCode) to override.  Then, if there is an extension that provides a custom implementation of toString() it would be the same as if the user wrote a custom toString() in the abstract template, causing AutoValue not to generate it's own implementation, right?